### PR TITLE
feat: Consume image asset metadata from CDK synth

### PR DESCRIPTION
**Pull request by mildaniel**
#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [inline samcli/lib/samlib/resource_metadata_normalizer.py:105]: Can we use Pathlib instead?

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [inline samcli/lib/samlib/resource_metadata_normalizer.py:107]: What happens in asset_path is `None`?

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [inline samcli/lib/samlib/resource_metadata_normalizer.py:90]: You should have tests that validate windows works as expected. Are Paths provided by the Metadata (aka CDK here) always paths that match the system? Meaning should we expect windows paths on windows or unix paths on windows?

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [inline samcli/lib/samlib/resource_metadata_normalizer.py:75]: I assume this is to find/update a nested value, like `Code.ImageUri`? If so, can we use `jmespath` instead? We use that lib in `package` to do similar things: https://github.com/aws/aws-sam-cli/blob/ec9f5d89971b823fa01071c96b7de90ea5e5303a/samcli/lib/package/artifact_exporter.py#L94

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [inline samcli/lib/samlib/resource_metadata_normalizer.py:49]: Why is the logical_id the `asset_path`?

**Review by jfuss — CHANGES_REQUESTED** [review]
Can we add integration tests for this, so we know how this behaves on Windows and Linux systems?

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (mildaniel) [reply samcli/lib/samlib/resource_metadata_normalizer.py:49]: This is equivalent to the name of the docker image generated by sam build and sets it to the value of Code.ImageUri.

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (mildaniel) [reply samcli/lib/samlib/resource_metadata_normalizer.py:75]: I tried using `set_value_from_jmespath()` but it might not work in this case. We get the template in with `Properties.Code = some_string` and it can't replace a string with a dict. Internally it recurses through the jmespath and sets the next property, so when it gets to the `Properties.Code` string it tries to do `some_string[key] = value` which obviously errors out.

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (mildaniel) [reply samcli/lib/samlib/resource_metadata_normalizer.py:107]: In theory it shouldn't ever be None, but if it is then the Lambda function image won't get built, and it'll fail when trying to local invoke.

**Comment by mildaniel** [comment]
> Can we add integration tests for this, so we know how this behaves on Windows and Linux systems?

We have an upcoming task for next sprint to add integration testing for all CDK-based build and local commands so I was deferring it until then. We don't currently have the underlying configurations in the develop branch to add integration tests with CDK.

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (mildaniel) [reply samcli/lib/samlib/resource_metadata_normalizer.py:90]: I did some messing around with CDK on Windows and it looks like `aws:asset:path` will return the system-matching paths, whereas `aws:asset:dockerfile-path` is always a unix-style relative path. I added a test for this Windows case.

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [reply samcli/lib/samlib/resource_metadata_normalizer.py:75]: I don't understand the explanation. Did you try to use it and it error-ed out?

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [inline samcli/lib/samlib/resource_metadata_normalizer.py:105]: Nit: follow [pep8](https://www.python.org/dev/peps/pep-0008/#imports). Specifically Absolute imports: `from pathlib import Path`

**Review by jfuss — CHANGES_REQUESTED** [review]

**Comment by jfuss** [comment]
> We have an upcoming task for next sprint to add integration testing for all CDK-based build and local commands so I was deferring it until then. We don't currently have the underlying configurations in the develop branch to add integration tests with CDK.

I don't understand why we are doing this. We can write integ tests now and assume the template format. We do this already but the same thing: https://github.com/aws/aws-sam-cli/blob/develop/tests/integration/local/invoke/test_integrations_cli.py#L67

I don't mean to be a strict here but I am not inclined to approve a pr without proper testing. We shouldn't be splitting testing from the implementation. We wouldn't accept this from contributors, why would we accept it for maintainers?

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (mildaniel) [reply samcli/lib/samlib/resource_metadata_normalizer.py:75]: Yes, that's correct.

**Comment by mildaniel** [comment]
@jfuss That's a completely fair call-out, I will add some integration testing.

> I don't understand why we are doing this. We can write integ tests now and assume the template format.

When writing this I was under the assumption that CDK integration tests would start from a CDK app -> `cdk synth` -> test CFN output. I see that we decided to go with testing starting from the CFN template which won't need some of the changes that would otherwise be required.

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (jfuss) [reply samcli/lib/samlib/resource_metadata_normalizer.py:77]: hmm ok. I won't block on this as it's small. I would like to get rid of it to make it easier to maintain but not the end of the world here.

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (jfuss) [inline tests/integration/local/invoke/invoke_cdk_templates.py:12]: Thank you!

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (jfuss) [inline tests/integration/local/invoke/invoke_cdk_templates.py:31]: This feels like we could abstract this into the base class but maybe there is a reason that won't work.

Why do this on `tearDownClass` over `teardown`? I expected to tear down after each function but maybe this is an optimization on tests? Whichever the reason, can you document it for clarity. Just helps us understand the thoughts behind things in the future.

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (jfuss) [inline tests/integration/local/invoke/invoke_cdk_templates.py:40]: Most of our docker tests are re-runs due to weird networking or docker issues: https://github.com/aws/aws-sam-cli/blob/develop/tests/integration/local/invoke/test_integrations_cli.py#L35 

Should we do the same here?

**Review by jfuss — CHANGES_REQUESTED** [review]
Just a couple small comments. Thanks for adding integ tests to this pr.

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (mildaniel) [reply tests/integration/local/invoke/invoke_cdk_templates.py:31]: This logic is inherent to this class. Not all classes that extend the base class need to be built and cleaned up. 

As for `tearDownClass` over `teardown`, we only need to run it once when we finish with these Docker images. Re-building and re-removing the images after each individual test will eat up a lot of unnecessary time since they don't change between tests.

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (mildaniel) [reply tests/integration/local/invoke/invoke_cdk_templates.py:40]: Looks like we also skip build tests on Windows in CI. Not exactly sure why but I'm assuming some Docker weirdness. Will add that in too.

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (jfuss) [reply tests/integration/local/invoke/invoke_cdk_templates.py:31]: That was my guess but would be good to just document that inline. Might be good for us to apply this logic across our tests. Flagging to @mndeveci and @CoshUS who I think are looking at some improvements in this area. Might not work in all cases but could be general useful for some things at least.

File: tests/integration/local/invoke/invoke_cdk_templates.py
Comment (jfuss) [reply tests/integration/local/invoke/invoke_cdk_templates.py:40]: We skip build tests because we run them separately: https://github.com/aws/aws-sam-cli/blob/develop/appveyor-windows-build-dotnet.yml

Might make sense to add these to the correct language windows appveyor job.

**Review by jfuss — APPROVED** [review]
Thanks for handling feedback and adding comments to make the code more readable!

**Review by wchengru — APPROVED** [review]

**Pull request merged**
